// @wix/ambassador-pro-gallery-v2-gallery/build/es/types.impl.js
export var Type;
(function (Type) {
  Type["UNDEFINED"] = "UNDEFINED";
  /** external link */
  Type["EXTERNAL"] = "EXTERNAL";
  /** for internal usage using wixLinkData */
  Type["INTERNAL"] = "INTERNAL";
})(Type || (Type = {}));
export var ItemType;
(function (ItemType) {
  ItemType["UNDEFINED"] = "UNDEFINED";
  ItemType["IMAGE"] = "IMAGE";
  ItemType["VIDEO"] = "VIDEO";
  ItemType["TEXT"] = "TEXT";
})(ItemType || (ItemType = {}));
export var ImageType;
(function (ImageType) {
  ImageType["UNDEFINED"] = "UNDEFINED";
  ImageType["WIX_MEDIA"] = "WIX_MEDIA";
  ImageType["EXTERNAL"] = "EXTERNAL";
})(ImageType || (ImageType = {}));
export var VideoType;
(function (VideoType) {
  VideoType["UNDEFINED"] = "UNDEFINED";
  VideoType["WIX_MEDIA"] = "WIX_MEDIA";
  VideoType["YOUTUBE"] = "YOUTUBE";
  VideoType["VIMEO"] = "VIMEO";
})(VideoType || (VideoType = {}));
export var Enum;
(function (Enum) {
  /** Default value. Means that permission not set */
  Enum["UNKNOWN"] = "UNKNOWN";
  /** Protected exposure. Exposed to members and owners */
  Enum["PROTECTED"] = "PROTECTED";
  /** Private exposure. Exposed to owners */
  Enum["PRIVATE"] = "PRIVATE";
  /** Public exposure. Visible to everyone */
  Enum["PUBLIC"] = "PUBLIC";
  /** Used for partial updates, to state that exposure is not changing */
  Enum["UNCHANGED"] = "UNCHANGED";
  /** Protected to members of permitted groups and owners */
  Enum["GROUP_PROTECTED"] = "GROUP_PROTECTED";
})(Enum || (Enum = {}));
export var SearchIndexingNotificationState;
(function (SearchIndexingNotificationState) {
  /** default state */
  SearchIndexingNotificationState["Unknown"] = "Unknown";
  /** metasite does not require site search indexing */
  SearchIndexingNotificationState["Off"] = "Off";
  /** metasite requires site search indexing */
  SearchIndexingNotificationState["On"] = "On";
})(SearchIndexingNotificationState || (SearchIndexingNotificationState = {}));
export var State;
(function (State) {
  State["UNDEFINED"] = "UNDEFINED";
  /** The gallery in the Editor segment */
  State["SAVED"] = "SAVED";
  /** The gallery in the LiveSite segment */
  State["PUBLISHED"] = "PUBLISHED";
})(State || (State = {}));
