// @wix/ambassador-members-routes-v1-route-configuration/build/es/http.impl.js
import { toURLSearchParams } from "@wix/metro-runtime";
import { serializer } from "@wix/metro-runtime/ambassador";
import { resolveUrl } from "@wix/metro-runtime";
var _bulkCreateRouteConfigurationRequest = {
  routeConfigurations: "_routeConfiguration",
};
var _bulkCreateRouteConfigurationResponse = {
  results: "_bulkRouteConfigurationResult",
};
var _bulkDeleteRouteConfigurationByWidgetIdRequest = {};
var _bulkDeleteRouteConfigurationByWidgetIdResponse = {
  results: "_bulkRouteConfigurationResult",
};
var _bulkDeleteRouteConfigurationRequest = {};
var _bulkDeleteRouteConfigurationResponse = {
  results: "_bulkRouteConfigurationResult",
};
var _bulkRouteConfigurationResult = { item: "_routeConfiguration" };
var _createRouteConfigurationRequest = {
  routeConfiguration: "_routeConfiguration",
};
var _createRouteConfigurationResponse = {
  routeConfiguration: "_routeConfiguration",
};
var _deleteRouteConfigurationByWidgetIdRequest = {};
var _deleteRouteConfigurationByWidgetIdResponse = {};
var _deleteRouteConfigurationRequest = {};
var _deleteRouteConfigurationResponse = {};
var _getRouteConfigurationByWidgetIdRequest = {};
var _getRouteConfigurationByWidgetIdResponse = {
  routeConfiguration: "_routeConfiguration",
};
var _getRouteConfigurationRequest = {};
var _getRouteConfigurationResponse = {
  routeConfiguration: "_routeConfiguration",
};
var _queryRouteConfigurationsRequest = {};
var _queryRouteConfigurationsResponse = {
  routeConfigurations: "_routeConfiguration",
};
var _reorderRouteConfigurationsRequest = {};
var _reorderRouteConfigurationsResponse = {
  routeConfigurations: "_routeConfiguration",
};
var _routeConfiguration = {
  createdDate: "google.protobuf.Timestamp",
  updatedDate: "google.protobuf.Timestamp",
};
var _updateRouteConfigurationByWidgetIdRequest = {
  fieldMask: "google.protobuf.FieldMask",
  routeConfiguration: "_routeConfiguration",
};
var _updateRouteConfigurationByWidgetIdResponse = {
  routeConfiguration: "_routeConfiguration",
};
var _updateRouteConfigurationRequest = {
  fieldMask: "google.protobuf.FieldMask",
  routeConfiguration: "_routeConfiguration",
};
var _updateRouteConfigurationResponse = {
  routeConfiguration: "_routeConfiguration",
};
function resolveWixMembersRoutesV1RouteConfigurationsUrl(opts) {
  var domainToMappings = {
    _: [
      {
        srcPath: "/_api/members/v1/route-configurations",
        destPath: "/v1",
      },
    ],
    "editor._base_domain_": [
      {
        srcPath: "/_api/members/v1/route-configurations",
        destPath: "/v1",
      },
    ],
    "blocks._base_domain_": [
      {
        srcPath: "/_api/members/v1/route-configurations",
        destPath: "/v1",
      },
    ],
    "create.editorx": [
      {
        srcPath: "/_api/members/v1/route-configurations",
        destPath: "/v1",
      },
    ],
    "members.wixapps.net": [
      {
        srcPath: "/_api/members/v1/route-configurations",
        destPath: "/v1",
      },
    ],
    "www._base_domain_": [
      {
        srcPath: "/_api/members/v1/route-configurations",
        destPath: "/v1",
      },
    ],
  };
  return resolveUrl(
    Object.assign(opts, { domainToMappings: domainToMappings })
  );
}
export function createRouteConfiguration(payload) {
  var _a = serializer(_createRouteConfigurationRequest, {
      _routeConfiguration: _routeConfiguration,
    }),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_createRouteConfigurationResponse, {
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __createRouteConfiguration(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "POST",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.CreateRouteConfiguration",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/route-configurations",
        data: serializedData,
        host: host,
      }),
      data: serializedData,
      transformResponse: fromRes,
    };
    return metadata;
  }
  __createRouteConfiguration.fromReq = fromReq;
  return __createRouteConfiguration;
}
createRouteConfiguration.__isAmbassador = true;
/** Creates up to 100 RouteConfigurations in a bulk. */
export function bulkCreateRouteConfiguration(payload) {
  var _a = serializer(_bulkCreateRouteConfigurationRequest, {
      _routeConfiguration: _routeConfiguration,
    }),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_bulkCreateRouteConfigurationResponse, {
    _bulkRouteConfigurationResult: _bulkRouteConfigurationResult,
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __bulkCreateRouteConfiguration(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "POST",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.BulkCreateRouteConfiguration",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/bulk/route-configurations/create",
        data: serializedData,
        host: host,
      }),
      data: serializedData,
      transformResponse: fromRes,
    };
    return metadata;
  }
  __bulkCreateRouteConfiguration.fromReq = fromReq;
  return __bulkCreateRouteConfiguration;
}
bulkCreateRouteConfiguration.__isAmbassador = true;
/** Retrieves a RouteConfiguration. */
export function getRouteConfiguration(payload) {
  var _a = serializer(_getRouteConfigurationRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_getRouteConfigurationResponse, {
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __getRouteConfiguration(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "GET",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.GetRouteConfiguration",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/route-configurations/{routeConfigurationId}",
        data: serializedData,
        host: host,
      }),
      params: toURLSearchParams(serializedData),
      transformResponse: fromRes,
    };
    return metadata;
  }
  __getRouteConfiguration.fromReq = fromReq;
  return __getRouteConfiguration;
}
getRouteConfiguration.__isAmbassador = true;
export function getRouteConfigurationByWidgetId(payload) {
  var _a = serializer(_getRouteConfigurationByWidgetIdRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_getRouteConfigurationByWidgetIdResponse, {
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __getRouteConfigurationByWidgetId(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "GET",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.GetRouteConfigurationByWidgetId",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath:
          "/v1/route-configurations/widget-id/{routeConfigurationWidgetId}",
        data: serializedData,
        host: host,
      }),
      params: toURLSearchParams(serializedData),
      transformResponse: fromRes,
    };
    return metadata;
  }
  __getRouteConfigurationByWidgetId.fromReq = fromReq;
  return __getRouteConfigurationByWidgetId;
}
getRouteConfigurationByWidgetId.__isAmbassador = true;
/**
 * Retrieves a list of RouteConfigurations, given the provided [paging, filtering, and sorting][1].
 * Default sorting - by "order" field Ascending
 * If sorting is not provided - applies default sorting.
 * Up to 1,000 RouteConfigurations can be returned per request.
 * To learn how to query RouteConfigurations, see [API Query Language][2].
 * [1]: https://dev.wix.com/api/rest/getting-started/sorting-and-paging
 * [2]: https://dev.wix.com/api/rest/getting-started/api-query-language
 */
export function queryRouteConfigurations(payload) {
  var _a = serializer(_queryRouteConfigurationsRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_queryRouteConfigurationsResponse, {
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __queryRouteConfigurations(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "POST",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.QueryRouteConfigurations",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/route-configurations/query",
        data: serializedData,
        host: host,
      }),
      data: serializedData,
      transformResponse: fromRes,
    };
    return metadata;
  }
  __queryRouteConfigurations.fromReq = fromReq;
  return __queryRouteConfigurations;
}
queryRouteConfigurations.__isAmbassador = true;
/** Updates a RouteConfiguration. */
export function updateRouteConfiguration(payload) {
  var _a = serializer(_updateRouteConfigurationRequest, {
      _routeConfiguration: _routeConfiguration,
    }),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_updateRouteConfigurationResponse, {
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __updateRouteConfiguration(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "PATCH",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.UpdateRouteConfiguration",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/route-configurations/{routeConfiguration.id}",
        data: serializedData,
        host: host,
      }),
      data: serializedData,
      transformResponse: fromRes,
    };
    return metadata;
  }
  __updateRouteConfiguration.fromReq = fromReq;
  return __updateRouteConfiguration;
}
updateRouteConfiguration.__isAmbassador = true;
/** Updates a RouteConfiguration by WidgetId. */
export function updateRouteConfigurationByWidgetId(payload) {
  var _a = serializer(_updateRouteConfigurationByWidgetIdRequest, {
      _routeConfiguration: _routeConfiguration,
    }),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_updateRouteConfigurationByWidgetIdResponse, {
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __updateRouteConfigurationByWidgetId(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "PATCH",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.UpdateRouteConfigurationByWidgetId",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath:
          "/v1/route-configurations/widget-id/{routeConfiguration.widgetId}",
        data: serializedData,
        host: host,
      }),
      data: serializedData,
      transformResponse: fromRes,
    };
    return metadata;
  }
  __updateRouteConfigurationByWidgetId.fromReq = fromReq;
  return __updateRouteConfigurationByWidgetId;
}
updateRouteConfigurationByWidgetId.__isAmbassador = true;
/** Deletes a RouteConfiguration. */
export function deleteRouteConfiguration(payload) {
  var _a = serializer(_deleteRouteConfigurationRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_deleteRouteConfigurationResponse, {}).fromJSON;
  function __deleteRouteConfiguration(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "DELETE",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.DeleteRouteConfiguration",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/route-configurations/{routeConfigurationId}",
        data: serializedData,
        host: host,
      }),
      params: toURLSearchParams(serializedData),
      transformResponse: fromRes,
    };
    return metadata;
  }
  __deleteRouteConfiguration.fromReq = fromReq;
  return __deleteRouteConfiguration;
}
deleteRouteConfiguration.__isAmbassador = true;
export function deleteRouteConfigurationByWidgetId(payload) {
  var _a = serializer(_deleteRouteConfigurationByWidgetIdRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(
    _deleteRouteConfigurationByWidgetIdResponse,
    {}
  ).fromJSON;
  function __deleteRouteConfigurationByWidgetId(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "DELETE",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.DeleteRouteConfigurationByWidgetId",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath:
          "/v1/route-configurations/widget-id/{routeConfigurationWidgetId}",
        data: serializedData,
        host: host,
      }),
      params: toURLSearchParams(serializedData),
      transformResponse: fromRes,
    };
    return metadata;
  }
  __deleteRouteConfigurationByWidgetId.fromReq = fromReq;
  return __deleteRouteConfigurationByWidgetId;
}
deleteRouteConfigurationByWidgetId.__isAmbassador = true;
export function bulkDeleteRouteConfiguration(payload) {
  var _a = serializer(_bulkDeleteRouteConfigurationRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_bulkDeleteRouteConfigurationResponse, {
    _bulkRouteConfigurationResult: _bulkRouteConfigurationResult,
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __bulkDeleteRouteConfiguration(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "DELETE",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.BulkDeleteRouteConfiguration",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/bulk/route-configurations/delete",
        data: serializedData,
        host: host,
      }),
      params: toURLSearchParams(serializedData),
      transformResponse: fromRes,
    };
    return metadata;
  }
  __bulkDeleteRouteConfiguration.fromReq = fromReq;
  return __bulkDeleteRouteConfiguration;
}
bulkDeleteRouteConfiguration.__isAmbassador = true;
export function bulkDeleteRouteConfigurationByWidgetId(payload) {
  var _a = serializer(_bulkDeleteRouteConfigurationByWidgetIdRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_bulkDeleteRouteConfigurationByWidgetIdResponse, {
    _bulkRouteConfigurationResult: _bulkRouteConfigurationResult,
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __bulkDeleteRouteConfigurationByWidgetId(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "DELETE",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.BulkDeleteRouteConfigurationByWidgetId",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/bulk/route-configurations/delete-by-widget-id",
        data: serializedData,
        host: host,
      }),
      params: toURLSearchParams(serializedData),
      transformResponse: fromRes,
    };
    return metadata;
  }
  __bulkDeleteRouteConfigurationByWidgetId.fromReq = fromReq;
  return __bulkDeleteRouteConfigurationByWidgetId;
}
bulkDeleteRouteConfigurationByWidgetId.__isAmbassador = true;
export function reorderRouteConfigurations(payload) {
  var _a = serializer(_reorderRouteConfigurationsRequest, {}),
    toReq = _a.toJSON,
    fromReq = _a.fromJSON;
  var fromRes = serializer(_reorderRouteConfigurationsResponse, {
    _routeConfiguration: _routeConfiguration,
  }).fromJSON;
  function __reorderRouteConfigurations(_a) {
    var host = _a.host;
    var serializedData = toReq(payload);
    var metadata = {
      entityFqdn: "wix.members.routes.v1.route_configuration",
      method: "POST",
      methodFqn:
        "wix.members.routes.v1.RouteConfigurations.ReorderRouteConfigurations",
      url: resolveWixMembersRoutesV1RouteConfigurationsUrl({
        protoPath: "/v1/route-configurations/reorder",
        data: serializedData,
        host: host,
      }),
      data: serializedData,
      transformResponse: fromRes,
    };
    return metadata;
  }
  __reorderRouteConfigurations.fromReq = fromReq;
  return __reorderRouteConfigurations;
}
reorderRouteConfigurations.__isAmbassador = true;
