// @wix/app-definition-ids/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 }),
  (exports.WIX_FAQ =
    exports.WIX_SITE_SEARCH =
    exports.WIX_MUSIC =
    exports.BANDSINTOWN =
    exports.WIX_PORTFOLIO =
    exports.WIX_PRO_GALLERY =
    exports.WIX_PHOTO_ALBUMS =
    exports.WIX_HOTELS_RUNNER =
    exports.WIX_VIDEO =
    exports.WIX_GET_SUBSCRIBERS =
    exports.INSTAGRAM_FEED =
    exports.DEVIANTART_FEED =
    exports.WIX_CHAT =
    exports.WIX_RESTAURANTS_TABLE_RESERVATIONS =
    exports.WIX_RESTAURANTS_MENUS_NEW =
    exports.WIX_RESTAURANTS_MENUS =
    exports.WIX_RESTAURANTS_ORDERS_NEW =
    exports.WIX_RESTAURANTS_ORDERS =
    exports.WIX_RESERVATIONS =
    exports.NOTIFICATIONS =
    exports.META_SITE =
    exports.MEMBERS =
    exports.MEMBERS_AREA_V2 =
    exports.MEMBERS_AREA =
    exports.WIX_NEW_STORES =
    exports.WIX_STORES =
    exports.WIX_ECOM =
    exports.PRICING_PLANS =
    exports.WIX_FORUM =
    exports.WIX_BOOKINGS =
    exports.WIX_EVENTS =
    exports.WIX_BLOG =
    exports.WIX_HOTELS =
    exports.ADMIN_PAGES =
    exports.DASHBOARD_SALES =
    exports.DASHBOARD_SUGGESTIONS =
    exports.DASHBOARD_SETUP =
    exports.BUSINESS_DASHBOARD =
    exports.WIX_CODE_PLATFORM_VISIBILITY =
    exports.WIX_ACCESSIBILITY_WIZARD =
    exports.WIX_CODE_SITE_MONITORING =
    exports.WIX_CODE_SECRETS_MANAGER =
    exports.WIX_CODE_RELEASE_MANAGER =
    exports.CLOUD_DATA_SCHEMA_NOTIFIER =
    exports.OLD_BLOG =
    exports.SITE_MEMBERS =
    exports.DATA_BINDING =
    exports.WIX_CODE_DUPLEXER =
    exports.WIX_CODE_SITE_EXTENSION =
    exports.WIX_CODE =
      void 0),
  (exports.getAppDependencies =
    exports.FERA_REVIEWS =
    exports.MULTILINGUAL =
    exports.MEMBERS_ACCOUNT_INFO =
    exports.MEMBERS_ABOUT =
    exports.MY_SUBSCRIPTIONS =
    exports.MAP_PLUGIN =
    exports.WIX_REVIEWS =
    exports.WISHLIST_BLOCKS =
    exports.RESTAURANTS =
    exports.WIX_FILE_SHARE =
    exports.RICH_CONTENT =
    exports.WIX_GROUPS =
    exports.WIX_CHALLENGES =
    exports.WIX_FORMS =
      void 0),
  (exports.WIX_CODE = "wix-code"),
  (exports.WIX_CODE_SITE_EXTENSION = "CloudSiteExtension"),
  (exports.WIX_CODE_DUPLEXER = "af6d536e-0fc6-428b-a7f9-1d8cd6aa3703"),
  (exports.DATA_BINDING = "dataBinding"),
  (exports.SITE_MEMBERS = "SiteMembers"),
  (exports.OLD_BLOG = "79f391eb-7dfc-4adf-be6e-64434c4838d9"),
  (exports.CLOUD_DATA_SCHEMA_NOTIFIER = "f1fbeff1-05c3-4121-bdb5-75e74b6e9e51"),
  (exports.WIX_CODE_RELEASE_MANAGER = "8c41f3a6-47f7-40be-ad77-02a72cb2ddb3"),
  (exports.WIX_CODE_SECRETS_MANAGER = "4b10fcce-732d-4be3-9d46-801d271acda9"),
  (exports.WIX_CODE_SITE_MONITORING = "675bbcef-18d8-41f5-800e-131ec9e08762"),
  (exports.WIX_ACCESSIBILITY_WIZARD = "9699c03d-5c19-4a7c-a454-e8109a6e43fc"),
  (exports.WIX_CODE_PLATFORM_VISIBILITY =
    "8048f115-449b-4dab-b0e6-e09d5ffc8199"),
  (exports.BUSINESS_DASHBOARD = "a7597ab5-1ed4-458f-a5c1-5884d5e14281"),
  (exports.DASHBOARD_SETUP = "0750d046-d599-4cfc-9f6b-e9f4af169aa9"),
  (exports.DASHBOARD_SUGGESTIONS = "cead4297-02ee-4c0f-ae07-461f4ba71b3b"),
  (exports.DASHBOARD_SALES = "f95c72b1-84de-48ec-a61b-d96d61315ffe"),
  (exports.ADMIN_PAGES = "1514b01b-cb08-41b4-8582-0a88551ac769"),
  (exports.WIX_HOTELS = "135aad86-9125-6074-7346-29dc6a3c9bcf"),
  (exports.WIX_BLOG = "14bcded7-0066-7c35-14d7-466cb3f09103"),
  (exports.WIX_EVENTS = "140603ad-af8d-84a5-2c80-a0f60cb47351"),
  (exports.WIX_BOOKINGS = "13d21c63-b5ec-5912-8397-c3a5ddb27a97"),
  (exports.WIX_FORUM = "14724f35-6794-cd1a-0244-25fd138f9242"),
  (exports.PRICING_PLANS = "1522827f-c56c-a5c9-2ac9-00f9e6ae12d3"),
  (exports.WIX_ECOM = "1380b703-ce81-ff05-f115-39571d94dfcd"),
  (exports.WIX_STORES = "1380b703-ce81-ff05-f115-39571d94dfcd"),
  (exports.WIX_NEW_STORES = "215238eb-22a5-4c36-9e7b-e7c08025e04e"),
  (exports.MEMBERS_AREA = "14cc59bc-f0b7-15b8-e1c7-89ce41d0e0c9"),
  (exports.MEMBERS_AREA_V2 = "b976560c-3122-4351-878f-453f337b7245"),
  (exports.MEMBERS = "14dbefd2-01b4-fb61-32a7-3abd44da4908"),
  (exports.META_SITE = "22bef345-3c5b-4c18-b782-74d4085112ff"),
  (exports.NOTIFICATIONS = "14f25924-5664-31b2-9568-f9c5ed98c9b1"),
  (exports.WIX_RESERVATIONS = "1475ab65-206b-d79a-856d-fa10bdb479ea"),
  (exports.WIX_RESTAURANTS_ORDERS = "13e8d036-5516-6104-b456-c8466db39542"),
  (exports.WIX_RESTAURANTS_ORDERS_NEW = "9a5d83fd-8570-482e-81ab-cfa88942ee60"),
  (exports.WIX_RESTAURANTS_MENUS = "13c1402c-27f2-d4ab-7463-ee7c89e07578"),
  (exports.WIX_RESTAURANTS_MENUS_NEW = "b278a256-2757-4f19-9313-c05c783bec92"),
  (exports.WIX_RESTAURANTS_TABLE_RESERVATIONS =
    "f9c07de2-5341-40c6-b096-8eb39de391fb"),
  (exports.WIX_CHAT = "14517e1a-3ff0-af98-408e-2bd6953c36a2"),
  (exports.DEVIANTART_FEED = "7d297b79-baed-46d7-ac58-4bd68dcb70d0"),
  (exports.INSTAGRAM_FEED = "14635256-b183-1c71-a4d2-f55179b80e8a"),
  (exports.WIX_GET_SUBSCRIBERS = "1375baa8-8eca-5659-ce9d-455b2009250d"),
  (exports.WIX_VIDEO = "14409595-f076-4753-8303-9a86f9f71469"),
  (exports.WIX_HOTELS_RUNNER = "826670ba-3a6b-4157-ac72-7c4fca9ce220"),
  (exports.WIX_PHOTO_ALBUMS = "13ff8629-c1fc-e289-e81f-bc8c8968e9d6"),
  (exports.WIX_PRO_GALLERY = "14271d6f-ba62-d045-549b-ab972ae1f70e"),
  (exports.WIX_PORTFOLIO = "d90652a2-f5a1-4c7c-84c4-d4cdcc41f130"),
  (exports.BANDSINTOWN = "1405ef82-0ee0-65fb-88a1-2f172aa3573c"),
  (exports.WIX_MUSIC = "13bb5d67-1add-e770-a71f-001277e17c57"),
  (exports.WIX_SITE_SEARCH = "1484cb44-49cd-5b39-9681-75188ab429de"),
  (exports.WIX_FAQ = "14c92d28-031e-7910-c9a8-a670011e062d"),
  (exports.WIX_FORMS = "14ce1214-b278-a7e4-1373-00cebd1bef7c"),
  (exports.WIX_CHALLENGES = "2936472a-a1ed-4ae5-9f71-614313a9f4e7"),
  (exports.WIX_GROUPS = "148c2287-c669-d849-d153-463c7486a694"),
  (exports.RICH_CONTENT = "949cfcc9-2a3f-4a96-bd6c-11d9d82763a4"),
  (exports.WIX_FILE_SHARE = "1537b24e-29d1-6d8f-b8e1-d6860f2f70b9"),
  (exports.RESTAURANTS = "9142e27e-2e30-4948-ad8f-7e128939efc4"),
  (exports.WISHLIST_BLOCKS = "a97c0203-062a-4dd5-97cf-a90f9800a13a"),
  (exports.WIX_REVIEWS = "a95a5fce-e370-4402-9ce4-96956acc747d"),
  (exports.MAP_PLUGIN = "c5fedde3-63c7-4041-a5fb-c77bbe6308dc"),
  (exports.MY_SUBSCRIPTIONS = "2bef2abe-7abe-43da-889c-53c1500a328c"),
  (exports.MEMBERS_ABOUT = "14dbef06-cc42-5583-32a7-3abd44da4908"),
  (exports.MEMBERS_ACCOUNT_INFO = "14cffd81-5215-0a7f-22f8-074b0e2401fb"),
  (exports.MULTILINGUAL = "14d84998-ae09-1abf-c6fc-3f3cace5bf19"),
  (exports.FERA_REVIEWS = "c9064755-5cc8-4faf-a5a6-41a0150a6640");
var dependencies = {};
function getAppDependencies(a) {
  return dependencies[a] || [];
}
exports.getAppDependencies = getAppDependencies;
